Tests/MessagingPush/PushHandling/AutomaticPushClickedIntegrationTest.swift

```diff
@@ -173,41 +173,114 @@ class AutomaticPushClickedIntegrationTest: IntegrationTest {
      */
 
-    func test_givenMultiplePushClickHandlers_simulateFcmSdkSwizzlingBehavior_expectNoInfiniteLoop() {
+    func test_onPushAction_givenMultiplePushClickHandlers_simulateFcmSdkSwizzlingBehavior_expectNoInfiniteLoop() {
         let givenPush = PushNotificationStub.getPushNotSentFromCIO()
         let givenOtherPushHandler = PushEventHandlerMock()
         let givenPushClickAction = PushNotificationActionStub(push: givenPush, didClickOnPush: true)
 
         let expectOtherClickHandlerHandlesPush = expectation(description: "Other push handler should handle push.")
         expectOtherClickHandlerHandlesPush.expectedFulfillmentCount = 1 // the other push click handler should only be called once, indicating an infinite loop is not created.
         givenOtherPushHandler.onPushActionClosure = { _, onComplete in
+            expectOtherClickHandlerHandlesPush.fulfill()
+
             // Like the FCM SDK does, make a call back to the app's current `UNUserNotificationCenterDelegate` instance. We simulate this by performing a push click, again.
-            self.performPushAction(givenPushClickAction) {}
+            self.performPushAction(givenPushClickAction, withCompletionHandler: onComplete)
+        }
+        addOtherPushEventHandler(givenOtherPushHandler)
+
+        performPushAction(givenPushClickAction)
+
+        waitForExpectations()
 
+        pushClickHandler.assertDidNotHandlePushClick() // CIO SDK should not handle push.
+    }
+
+    func test_shouldDisplayPushAppInForeground_givenMultiplePushClickHandlers_simulateFcmSdkSwizzlingBehavior_expectNoInfiniteLoop() {
+        let givenPush = PushNotificationStub.getPushNotSentFromCIO()
+        let givenOtherPushHandler = PushEventHandlerMock()
+
+        let expectOtherClickHandlerHandlesPush = expectation(description: "Other push handler should handle push.")
+        expectOtherClickHandlerHandlesPush.expectedFulfillmentCount = 1 // the other push click handler should only be called once, indicating an infinite loop is not created.
+        givenOtherPushHandler.shouldDisplayPushAppInForegroundClosure = { _, onComplete in
             expectOtherClickHandlerHandlesPush.fulfill()
-            onComplete()
+
+            // Like the FCM SDK does, make a call back to the app's current `UNUserNotificationCenterDelegate` instance. We simulate this by sending the push event back to us.
+            self.sendPushEventShouldShowPushAppInForeground(givenPush, withCompletionHandler: onComplete)
         }
         addOtherPushEventHandler(givenOtherPushHandler)
 
-        performPushAction(givenPushClickAction)
+        sendPushEventShouldShowPushAppInForeground(givenPush)
 
         waitForExpectations()
 
         pushClickHandler.assertDidNotHandlePushClick() // CIO SDK should not handle push.
     }
 
+    /*
```

Comment:
While QA testing, this is a new scenario that I encountered: 3rd party SDK calling the completionHandler 2+ times. 

These are automated tests that reproduce what I encountered during QA testing.